1.8.0-rc.0: unable to compile with supplied Makefile

Hi all,

Having a quick look at 1.8 rc 0 and could not compile it with the supplied Makefile.

### Environment

Erlang 21.2-1 (esl-erlang from Erlang Solns ppa)
 
Ubuntu 18.04 with latest maintenance

### Current behaviour

Standard elixir build:

`make clean test`

and I eventually see some warnings (which seem to be in 1.7.4 also):

> ==> elixir (compile)
> warning: module Exception is not a behaviour (in module IO.StreamError)
  lib/io/stream.ex:1

> warning: got "@impl true" for function message/1 but no behaviour specifies such callback. There are no known callbacks, please specify the proper @behaviour and make sure it defines callbacks
  lib/io/stream.ex:2

> warning: got "@impl true" for function exception/1 but no behaviour specifies such callback. There are no known callbacks, please specify the proper @behaviour and make sure it defines callbacks
  lib/io/stream.ex:2

> warning: got "@impl true" for function exception/1 but no behaviour specifies such callback. There are no known callbacks, please specify the proper @behaviour and make sure it defines callbacks
  lib/io/stream.ex:2

> warning: got "@impl true" for function exception/1 but no behaviour specifies such callback. There are no known callbacks, please specify the proper @behaviour and make sure it defines callbacks
  lib/io/stream.ex:5

and then these error messages:

`Compilation failed due to warnings while using the --warnings-as-errors option`
`Makefile:74: recipe for target 'lib/elixir/ebin/Elixir.Kernel.beam' failed`
`make: *** [lib/elixir/ebin/Elixir.Kernel.beam] Error 1`

### Expected behavior

If I remove the `+warnings-as_errors` from the `ERLC` in the Makefile, the make completes with all tests passing.

The `+warnings_as_errors` is not in the Makefile for on 1.7.4. I can build 1.7.4 no problem.

Couldn't' see anything in the CHANGELOG that would throw any light on this. Just me?

--- a/lib/elixir/lib/kernel/typespec.ex
+++ b/lib/elixir/lib/kernel/typespec.ex
@@ -567,11 +567,12 @@ defmodule Kernel.Typespec do
   end
 
   # Handle records
-  defp typespec({:record, meta, [atom]}, vars, caller, state) when is_atom(atom) do
+  defp typespec({:record, meta, [atom]}, vars, caller, state) do
     typespec({:record, meta, [atom, []]}, vars, caller, state)
   end
 
-  defp typespec({:record, meta, [tag, field_specs]}, vars, caller, state) when is_atom(tag) do
+  defp typespec({:record, meta, [tag, field_specs]}, vars, caller, state)
+       when is_atom(tag) and is_list(field_specs) do
     # We cannot set a function name to avoid tracking
     # as a compile time dependency because for records it actually is one.
     case Macro.expand({tag, [], [{:{}, [], []}]}, caller) do
@@ -594,8 +595,9 @@ defmodule Kernel.Typespec do
     end
   end
 
-  defp typespec({:record, _meta, _args}, _vars, caller, _state) do
-    compile_error(caller, "invalid record specification, expected the record name to be an atom")
+  defp typespec({:record, _meta, [_tag, _field_specs]}, _vars, caller, _state) do
+    message = "invalid record specification, expected the record name to be an atom literal"
+    compile_error(caller, message)
   end
 
   # Handle ranges

--- a/lib/elixir/test/elixir/typespec_test.exs
+++ b/lib/elixir/test/elixir/typespec_test.exs
@@ -571,12 +571,23 @@ defmodule TypespecTest do
       end
     end
 
-    test "@type with invalid record" do
-      assert_raise CompileError, ~r"invalid record specification", fn ->
+    test "@type with a record which declares the name as the type `atom` rather than an atom literal" do
+      assert_raise CompileError, ~r"expected the record name to be an atom literal", fn ->
         test_module do
-          @type my_type :: record(atom)
+          @type my_type :: record(atom, field: :foo)
         end
       end
+    end
+
+    test "@type can be named record" do
+      bytecode =
+        test_module do
+          @type record :: binary
+          @spec foo?(record) :: boolean
+          def foo?(_), do: true
+        end
+
+      assert [type: {:record, {:type, _, :binary, []}, []}] = types(bytecode)
     end
 
     test "@type with an invalid map notation" do

